File: automo-client-py/connection.py
import json
import requests
import datetime
import dateutil.relativedelta
from getpass import getpass
from pprint import pprint

from user import User

class Connection:
    user = None
    index_url = None
    index_data = {}


    def login(self, index_url, username, password):
        user = User()
        self.index_url = index_url
        if not user.login(index_url, username, password):
            return False

        self.user = user
        return True


    def is_loggedin(self):
        if self.user is None:
            print("Connection: No Login User")
            return False

        if not self.user.is_loggedin():
            print("Connection: Not Logged In")
            return False

        return True

    
    def get_index(self):
        if not self.index_data:
            self.index_data = self.get(self.index_url)
        return self.index_data


    def get(self, url, params=None):
        if not self.is_loggedin():
            return None

        auth_retry = 3
        while auth_retry > 0:
            try:
                r = requests.get(url, params=params, auth=self.user.get_auth_params())
            except requests.ConnectionError:
                print("Get: Connection Error")
                return False

            if r.status_code == 401:
                print("Get: Not Authorized")
                auth_retry -= 1
                if not self.user.get_token():
                    break
            else:
                break

        if r.status_code != 200:
            print("Get: Response Error {}".format(r.status_code))
            print(r.text)
            return None

        try:
            data = json.loads(r.text)
        except json.JSONDecodeError:
            print("Get: JSON decode error")
            print(data)
            return None
        
        return data


    def post_json(self, url, data, params=None):
        if not self.is_loggedin():
            return None

        auth_retry = 3
        while auth_retry > 0:
            try:
                #r = requests.get(url, params=params, auth=self.user.get_auth_params())
                r = requests.post(url, json=data, params=params, auth=self.user.get_auth_params())
            except requests.ConnectionError:
                print("Post: Connection Error")
                return False

            if r.status_code == 401:
                print("Post: Not Authorized")
                auth_retry -= 1
                if not self.user.get_token():
                    break
            else:
                break

        if r.status_code != 200:
            print("Post: Response Error {}".format(r.status_code))

        try:
            response_data = json.loads(r.text)
        except json.JSONDecodeError:
            print("Post: JSON decode error")
            print(r.text)
            return None
        
        return response_data

File: automo-client-js/src/app/connection.js
const status = require("./status")
const User = require("./user")


module.exports = class Connection {
    constructor(logger) {
        this.index_url = null
        this.resource_index = {}
        this.user = null;
        this.logger = logger;
    }


    login(index_url, username, password, on_success, on_failed, on_finally) {
        this.index_url = index_url;
        this.user = new User();
        this.user.login(
            index_url,
            username,
            password,
            (resource_index) => {
                this.resource_index = resource_index;
                on_success();
                on_finally != null ? on_finally() : false;
            },
            (error) => {
                on_failed(error);
                on_finally != null ? on_finally() : false;
            }
        )
    }

    resourceFromPath(pathList) {
        var result = this.resource_index
        pathList.forEach((key) => {
            result = result[key]
        })
        return result
    }


    logout(on_success, on_failed) {
        this.user = null;
        on_success();
    }


    isLoggedIn() {
        if (this.user === null) {
            return false;
        }
        if (!(this.user.tokenValid())) {
            return false;
        }
        return true
    }


    _authenticate_and_execute(func, on_failed, on_finally) {
        if (this.user == null) {
            on_failed(new Error("User not logged in"));
            on_finally != null ? on_finally() : false;
            return;
        }
        if (!this.user.tokenValid()) {
            this.user.getToken(
                () => {
                    func()
                },
                (error) => {
                    on_failed(error)
                    on_finally != null ? on_finally() : false;
                }
            );
            return;
        }
        func()
    }


    _get(url, on_success, on_failed, on_finally, refetchTokenOnFail = true, method='GET') {
        let headers = this.user.getAuthorizationHeaders();

        fetch(url, { method: method, headers: headers })
            .then(status)
            .then(response => response.json())
            .then(data => {
                on_success(data);
                on_finally != null ? on_finally() : false;
            })
            .catch(error => {
                if (refetchTokenOnFail ? (error.status == 401) : false) {
                    this.user.getToken(
                        () => {
                            this._get(url, on_success, on_failed, on_finally, false)
                        },
                        (getTokenError) => {
                            on_failed(getTokenError);
                            on_finally != null ? on_finally() : false;
                        }
                    );
                } else {
                    on_failed(error);
                    on_finally != null ? on_finally() : false;
                }
            })
    }


    get(url, on_success, on_failed, on_finally) {
        this._authenticate_and_execute(
            () => {
                this._get(url, on_success, on_failed, on_finally)
            },
            on_failed,
            on_finally
        )
    }


    delete(url, on_success, on_failed, on_finally) {
        this._authenticate_and_execute(
            () => {
                this._get(url, on_success, on_failed, on_finally, true, 'DELETE')
            },
            on_failed,
            on_finally
        )
    }


    _get_blob(url, on_success, on_failed, on_finally, refetchTokenOnFail = true) {
        let headers = this.user.getAuthorizationHeaders();

        fetch(url, { method: 'GET', headers: headers })
            .then(status)
            .then(response => response.blob())
            .then(blob => {
                on_success(blob);
                on_finally != null ? on_finally() : false;
            })
            .catch(error => {
                if (refetchTokenOnFail ? (error.status == 401) : false) {
                    this.user.getToken(
                        () => {
                            this._get(url, on_success, on_failed, on_finally, false)
                        },
                        (getTokenError) => {
                            on_failed(getTokenError);
                            on_finally != null ? on_finally() : false;
                        }
                    );
                } else {
                    on_failed(error);
                    on_finally != null ? on_finally() : false;
                }
            })
    }


    get_blob(url, on_success, on_failed, on_finally) {
        this.logger.log_spinner(`GET ${url}...`)
        this._authenticate_and_execute(
            () => {
                this._get_blob(url, on_success, on_failed, on_finally);
            },
            on_failed,
            on_finally
        )
    }

    _post(url, post_data, on_success, on_failed, on_finally, refetchTokenOnFail = true) {
        let headers = this.user.getAuthorizationHeaders();

        headers.set('Content-Type', 'application/json');

        fetch(url, { method: 'POST', body: JSON.stringify(post_data), headers: headers })
            .then(status)
            .then(response => response.json())
            .then(data => {
                on_success(data);
                on_finally != null ? on_finally() : false;
            })
            .catch(error => {
                if (refetchTokenOnFail ? (error.status == 401) : false) {
                    this.user.getToken(
                        () => {
                            this._post(url, post_data, on_success, on_failed, on_finally, false)
                        },
                        (getTokenError) => {
                            on_failed(getTokenError);
                            on_finally != null ? on_finally() : false;
                        }
                    );
                } else {
                    on_failed(error);
                    on_finally != null ? on_finally() : false;
                }
            })
    }


    post(url, post_data, on_success, on_failed, on_finally) {
        this._authenticate_and_execute(
            () => {
                this._post(url, post_data, on_success, on_failed, on_finally);
            },
            on_failed,
            on_finally
        )
    }
}